compiler/src/iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp

```diff
@@ -368,6 +405,138 @@ void LinalgFusePass::runOnOperation() {
     return signalPassFailure();
   }
 }
+namespace {
+/// Pattern to wrap splitReduction transformation.
+struct CodegenSplitReduction
+    : public OpInterfaceRewritePattern<linalg::LinalgOp> {
+  CodegenSplitReduction(MLIRContext *context, bool reductionReordering,
+                        int64_t size, linalg::LinalgTransformationFilter filter,
+                        PatternBenefit benefit = 1)
+      : OpInterfaceRewritePattern<linalg::LinalgOp>(context, benefit),
+        reductionReordering(reductionReordering),
+        size(size),
+        filter(std::move(filter)) {}
+
+  LogicalResult matchAndRewrite(linalg::LinalgOp op,
+                                PatternRewriter &rewriter) const override {
+    // Make sure that
+    // a) The pass was not applied earlier
+    // b) operation has tensor semantics
+    // c) number of reduction loops == 1
+    // d) operation only has projection permutations for index map
+    // e) It's a linalg generic op
+    // f) it has only 1 input
+    // g) if enableReductionReordering is not set, then operand is an int
+    // h) reduction dimension is the innermost dimension of the operand shape
+    // TODO: support named ops, numInputs > 1, and modify lastDim check below
+    // accordingly. If reductionReordering is not enabled by default, it must be
+    // an integer or index type to proceed to allow associative reordering.
+    if (failed(filter.checkAndNotify(rewriter, op)) ||
+        !op.hasTensorSemantics() || op.getNumReductionLoops() != 1 ||
+        op.getNumOutputs() != 1 || !op.hasOnlyProjectedPermutations() ||
+        !isa<linalg::GenericOp>(op) || op.getNumInputs() != 1 ||
+        !(reductionReordering || op.getInputOperand(0)
+                                     ->get()
+                                     .getType()
+                                     .dyn_cast<ShapedType>()
+                                     .getElementType()
+                                     .isIntOrIndex()))
+      return rewriter.notifyMatchFailure(
+          op,
+          "pass was applied earlier or "
+          "operation doesn't have tensor semantics or "
+          "number of reduction loops != 1 or "
+          "operation has non projected permutations for index map or "
+          "it's not a linalg generic op or "
+          "it has multiple inputs or "
+          "if enableReductionReordering is not set and "
+          "operand is not an int");
```

Comment:
Please, split these checks into individual ones so that the condition per check is simpler and we can return a single specific error message.